src/challenge.rs

```diff
@@ -0,0 +1,66 @@
+use ark_ff::PrimeField;
+use ark_sponge::CryptographicSponge;
+
+/// Challenge Generator (todo doc)
+/// Note that mutable reference cannot be cloned.
+#[derive(Clone)]
+pub enum ChallengeGenerator<F: PrimeField, S: CryptographicSponge> {
+    /// Each challenge is freshly squeezed from a sponge.
+    Multivariate(S),
+    /// Each challenge is a power of one squeezed element from sponge.
+    ///
+    /// `Univariate(generator, next_element)`
+    Univariate(F, F),
+}
+
+impl<F: PrimeField, S: CryptographicSponge> ChallengeGenerator<F, S> {
+    /// Returns a challenge generator with multivariate strategy. Each challenge is freshly squeezed
+    /// from a sponge.
+    pub fn new_multivariate(sponge: S) -> Self {
+        Self::Multivariate(sponge)
+    }
+
+    /// Returns a challenge generator with univariate strategy. Each challenge is a power of one
+    /// squeezed element from sponge.
+    pub fn new_univariate(sponge: &mut S) -> Self {
+        let gen = sponge.squeeze_field_elements(1)[0];
+        Self::Univariate(gen, gen)
+    }
+
+    /// Returns the next challenge generated.
+    pub fn next_challenge(&mut self) -> F {
+        match self {
+            Self::Multivariate(s) => s.squeeze_field_elements(1)[0],
+            Self::Univariate(gen, next) => {
+                let result = next.clone();
+                *next *= *gen;
+                result
+            }
+        }
+    }
+
+    /// Returns the next challenge generated where next challenge has `size` bits. Only works for
+    /// multivariate generator.
+    ///
+    /// ## Panics
+    /// This function will panic if `self` is univariate.
+    pub fn next_challenge_of_size(&mut self, size: ark_sponge::FieldElementSize) -> F {
+        match self {
+            Self::Multivariate(s) => s.squeeze_field_elements_with_sizes(&[size])[0],
+            _ => {
+                panic!("`next_challenge_of_size` only supports multivariate generator.")
+            }
+        }
+    }
```

Comment:
I think we shouldn't panic here? Because `poly-commit` is now in charge of challenges, we have to hardcode (in the implementation of the PC scheme) whether to invoke `next_challenge_of_size` or `next_challenge`. (In contrast, earlier we passed in a closure that we defined.)

Maybe this indicates that we should go back to the closure method, so that the caller has flexibility in generation? Alternatively, we can set `Multivariate` to also contain a "challenge_size: ChallengeSizeInBits" field which is `Full` by default. `next_challenge` inspects this field and uses it to determine what kind of challenge to generate. Does that work better?

What do you think @weikengchen?

Reply:
Any suggestions for this? This is the last blocker for this PR

Reply:
Setting `Multivariate` to contain `challenge_size` looks good to me. @weikengchen any input on that? If it sounds good I will do that tonight

Reply:
I think that is a good solution.

Reply:
Will do it tonight

Reply:
fixed